src/main/scala/is/hail/utils/BinaryHeap.scala

```diff
@@ -0,0 +1,154 @@
+package is.hail.utils
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+class BinaryHeap[@specialized T : ClassTag](initialCapacity: Int = 32) {
+  private class RankedT(val v: T, var rank: Long) {}
+  private var a: Array[RankedT] = new Array[RankedT](initialCapacity)
```

Comment:
I thought about this, and I'd want to add random access writing to ArrayBuilder, @jigold @cseed  thoughts on this change? I also like my shrinking strategy 😉, but I could go with ArrayBuilder if shrinking is not supported by practical experience.